src/ompl/geometric/planners/experience/src/LRR.cpp

```diff
@@ -0,0 +1,610 @@
+ *   * Neither the name of the Univ of CO, Boulder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Dave Coleman */
+
+#include "ompl/geometric/planners/experience/LRR.h"
+#include "ompl/geometric/planners/rrt/RRTConnect.h"
+#include "ompl/base/goals/GoalState.h"
+#include "ompl/base/goals/GoalSampleableRegion.h"
+#include "ompl/tools/config/SelfConfig.h"
+#include "ompl/tools/lightning/LightningDB.h"
+
+#include <boost/thread.hpp>
+
+#include <limits>
+
+ompl::geometric::LRR::LRR(const base::SpaceInformationPtr &si, const ompl::tools::LightningDBPtr &experienceDB)
+    : base::Planner(si, "LRR"),
+      experienceDB_(experienceDB),
+      nearestK_(10) // default value
+{
+    specs_.approximateSolutions = true;
+    specs_.directed = true;
+
+    // Repair Planner Specific:
+    repairProblemDef_.reset(new base::ProblemDefinition(si_));
+
+    psk_.reset(new ompl::geometric::PathSimplifier(si_));
+}
+
+ompl::geometric::LRR::~LRR(void)
+{
+    freeMemory();
+}
+
+void ompl::geometric::LRR::clear(void)
+{
+    Planner::clear();
+    freeMemory();
+
+    // Clear the inner planner
+    if (repairPlanner_)
+        repairPlanner_->clear();
+}
+
+void ompl::geometric::LRR::setLightningDB(ompl::tools::LightningDBPtr experienceDB)
+{
+    experienceDB_ = experienceDB;
+}
+
+void ompl::geometric::LRR::setRepairPlanner(const base::PlannerPtr &planner)
+{
+    if (planner && planner->getSpaceInformation().get() != si_.get())
+        throw Exception("Repair planner instance does not match space information");
+    repairPlanner_ = planner;
+    setup_ = false;
+}
+
+void ompl::geometric::LRR::setup(void)
+{
+    Planner::setup();
+
+    // Setup repair planner (for use by the rrPlanner)
+    // Note: does not use the same pdef as the main planner in this class
+    if (!repairPlanner_)
+    {
+        // Set the repair planner
+        boost::shared_ptr<ompl::geometric::RRTConnect> repair_planner( new ompl::geometric::RRTConnect( si_ ) );
```

Comment:
I'm not sure that is a good idea - I believe our repair problem does not have a goal samplable region and thus it would never choose RRTConnect which is what the paper recommends. I don't think its sampleable because I just take the state before and after an invalid region on a path and try to replan between them.